### PR TITLE
Move secret validation logic and authN-loadPrincipal into `PolarisSecretsManager`

### PR DESCRIPTION
**Pull request by snazy**
The logic _how_ a principal and/or principal secret's are persisted should be transparent to the calling code. Relying on the persistence internals for principals and secrets management makes it impossible to factor out secrets management / make principal management possible.

This change moves the secret validation and retrieval of a principal by client-ID behind an implementation of `PolarisSecretsManager`.

### CODE REVIEW

File: polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecretsManager.java [hunk 1]
```diff
@@ -39,6 +41,17 @@ public interface PolarisSecretsManager {
...
+  @Nonnull
+  SecretValidationResult validateSecret(
+      @Nonnull PolarisCallContext callCtx, @Nonnull String clientId, @Nonnull String clientSecret);
+
+  @Nonnull
+  EntityResult loadPrincipal(
+      @Nonnull PolarisCallContext callCtx,
+      @Nullable String roleName,
+      @Nullable String clientId,
+      @Nullable Long principalId);
```
Comment (collado-mike) [inline polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecretsManager.java:53]: Loading the principal entity should be the purview of the persistence layer, not the secrets manager. IMO, the secrets manager should be something like Vault or AWS Secrets manager that knows nothing about Polaris entities, but does know how to store client id / secret

File: polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecretsManager.java [hunk 2]
```diff
@@ -100,4 +113,34 @@ public PolarisPrincipalSecrets getPrincipalSecrets() {
...
+        @JsonProperty("returnStatus") @Nonnull BaseResult.ReturnStatus returnStatus,
+        @JsonProperty("extraInformation") @Nullable String extraInformation,
+        @JsonProperty("principalSecrets") @Nonnull PolarisBaseEntity principal) {
+      super(returnStatus, extraInformation);
+      this.principal = principal;
+    }
+
+    public PolarisBaseEntity getPrincipal() {
+      return principal;
+    }
```
Comment (collado-mike) [inline polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecretsManager.java:144]: Same comment on returning the principal from the secret validation command. I can imagine storing extra metadata, such as the principal id with the client id / secret, but the entity itself should be returned from the persistence layer.

File: polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecretsManager.java (hunk 1 above)
Comment (snazy) [reply polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecretsManager.java:53]: The problem is: effectively only the secrets manager knows which attributes it has at hand, correct?
Currently, it's the principal-ID, which is another unique ID in the JWT - but it should rather be just the client-ID. But that's currently hard to refactor, because the data model details leak all the way up. That's why #512 is there as an immediate follow-up.
TL;DR this PR is one of a series of upcoming PRs to untangle the hard dependency of all the services on the data model details/internals.

Comment (collado-mike) [reply polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecretsManager.java:53]: Vault, Secrets Manager, and K8s all allow storing secret values as structured types, like JSON. I think it's pretty easy to imagine a layout like using the `clientId` as the secret key and storing the secret value as something like `{"secret": "abcd", "principalId": 1}`. This allows full decoupling of the PolarisSecretsManager from the PolarisMetaStoreManager, whereas returning the `PrincipalEntity` from the `PolarisSecretsManager` means it has to maintain a reference to the MetaStore and the MetaStore has to deal with lookup keys beside name and id. No other entity has a lookup key beside name or id, so I think keeping it consistent for Principal makes sense.

Comment (snazy) [reply polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecretsManager.java:53]: The issue here is: there are currently _two_ lookup keys: client-ID _and_ principal-ID. Client-ID is the natural key, so I think that we should only use that one.

Comment (snazy) [reply polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecretsManager.java:53]: I think, the assumption that Polaris is _the_ source-of-truth for principals is not correct. Principals/identities are usually managed elsewhere. It would be difficult to force people to add a principal-ID as a new attribute to new or existing identities. On top, principal-ID is an internal concern.

Comment (collado-mike) [reply polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecretsManager.java:53]: The secrets manager is only used for Polaris-managed Principals. There's no need to manage secrets for externally managed Principals. I think the assumption that Polaris is authoritative here is a safe one
